src/extensions/agent_modules/shared/enrollment_utils/src/adu_enrollment_utils.c

```diff
@@ -0,0 +1,126 @@
+#include "aduc/adu_enrollment_utils.h"
+#include <aduc/agent_state_store.h>
+#include <aduc/logging.h>
+#include <aduc/string_c_utils.h>
+#include <string.h> // strncpy
+
+/**
+ * @brief Gets the enrollment data object from the enrollment operation context.
+ * @param context The enrollment operation context.
+ * @return ADUC_Enrollment_Request_Operation_Data* The enrollment data object.
+ */
+ADUC_Enrollment_Request_Operation_Data* EnrollmentData_FromOperationContext(ADUC_Retriable_Operation_Context* context)
+{
+    if (context == NULL || context->data == NULL)
+    {
+        Log_Error("Null input (context:%p, data:%p)", context, context->data);
+        return NULL;
+    }
+
+    return (ADUC_Enrollment_Request_Operation_Data*)(context->data);
+}
+
+/**
+ * @brief Sets the enrollment state and updates 'IsDeviceEnrolled' state in the DU agent state store.
+ * @param data The enrollment data object.
+ * @param state The enrollment state to set.
+ * @return ADU_ENROLLMENT_STATE The old enrollment state.
+ */
+ADU_ENROLLMENT_STATE EnrollmentData_SetState(
+    ADUC_Enrollment_Request_Operation_Data* enrollmentData, ADU_ENROLLMENT_STATE state, const char* reason)
+{
+    ADU_ENROLLMENT_STATE oldState = enrollmentData->enrollmentState;
+    if (enrollmentData->enrollmentState != state)
+    {
+        Log_Info(
+            "Enrollment state changed from %d to %d (reason:%s)",
+            enrollmentData->enrollmentState,
+            state,
+            IsNullOrEmpty(reason) ? "unknown" : reason);
+        enrollmentData->enrollmentState = state;
+
+        if (ADUC_StateStore_SetIsDeviceEnrolled(state == ADU_ENROLLMENT_STATE_ENROLLED) != ADUC_STATE_STORE_RESULT_OK)
+        {
+            Log_Error("Failed to set enrollment state in state store");
+        }
+    }
+    return oldState;
+}
+
+/**
+ * @brief Sets the correlation id for the enrollment request message.
+ * @param enrollmentData The enrollment data object.
+ * @param correlationId The correlation id to set.
+ */
+void EnrollmentData_SetCorrelationId(ADUC_Enrollment_Request_Operation_Data* enrollmentData, const char* correlationId)
+{
+    if (enrollmentData == NULL || correlationId == NULL)
+    {
+        Log_Error("Null input (enrollmentData:%p, correlationId:%p)", enrollmentData, correlationId);
+        return;
+    }
+
+    strncpy(
```

Comment:
don't we have "safe" version  of this in our string library?

Reply:
unfortunately no.  I will send a PR on develop branch that creates a banned.h that does `#undef strncpy` and `#define strncpy(s, t, n) strncpy_is_BANNED` or also create a macro to define for compilation error for any banned functions.

I suppose it would have to do `#include <banned.h>` in every compilation unit and we can have a githook that enforced every .c or .cpp includes it. Maybe there's a better way?

Reply:
Fixed in develop with PR https://github.com/Azure/iot-hub-device-update/pull/560
Once that is merged, will cherry-pick to gen2 topic branch and use ADUC_Safe_StrCopyN replacement instead of strncpy here and will ban it by #include of aduc_banned.h